ScheduleTask panic when there is no executor

### What did you do?

Run a single tiflow master, and then create a job.

### What did you expect to see?

tiflow master should be running as well.

### What did you see instead?

```
panic: invalid argument to Intn

goroutine 484 [running]:
math/rand.(*Rand).Intn(0xc001dcf580?, 0x5d574b0?)
	/home/xyj/.local/lib/go1.19.2/src/math/rand/rand.go:168 +0x65
math/rand.Intn(...)
	/home/xyj/.local/lib/go1.19.2/src/math/rand/rand.go:337
github.com/pingcap/tiflow/engine/servermaster/scheduler.(*Scheduler).ScheduleTask(0xc000d523f0?, {0x5d574b0?, 0xc002017740?}, 0x50cf040?)
	/home/xyj/Projects/tiflow/engine/servermaster/scheduler/scheduler.go:56 +0x46
github.com/pingcap/tiflow/engine/servermaster.(*Server).ScheduleTask(0xc0005b3560, {0x5d574b0, 0xc002017740}, 0x37?)
	/home/xyj/Projects/tiflow/engine/servermaster/server.go:318 +0x108
github.com/pingcap/tiflow/engine/enginepb._TaskScheduler_ScheduleTask_Handler.func1({0x5d574b0, 0xc002017740}, {0x50cf040?, 0xc000d523f0})
	/home/xyj/Projects/tiflow/engine/enginepb/master_grpc.pb.go:431 +0x78
github.com/pingcap/tiflow/engine/pkg/rpcutil.UnaryServerInterceptor({0x5d574b0?, 0xc002017740?}, {0x50cf040, 0xc000d523f0}, 0xc0003fb5a0, 0xc000d24f50?)
	/home/xyj/Projects/tiflow/engine/pkg/rpcutil/middleware.go:100 +0x5a
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x5d574b0?, 0xc002017740?}, {0x50cf040?, 0xc000d523f0?})
	/home/xyj/.local/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1129 +0x5b
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1({0x5d574b0, 0xc002017740}, {0x50cf040, 0xc000d523f0}, 0x50e6480?, 0xc00203e300)
	/home/xyj/.local/go/pkg/mod/github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0x87
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x5d574b0?, 0xc002017740?}, {0x50cf040?, 0xc000d523f0?})
	/home/xyj/.local/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1132 +0x83
google.golang.org/grpc.chainUnaryInterceptors.func1({0x5d574b0, 0xc002017740}, {0x50cf040, 0xc000d523f0}, 0xc0003fb5a0, 0xc000d239c8)
	/home/xyj/.local/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1134 +0x12b
github.com/pingcap/tiflow/engine/enginepb._TaskScheduler_ScheduleTask_Handler({0x52c93a0?, 0xc0005b3560}, {0x5d574b0, 0xc002017740}, 0xc000d52380, 0xc001906620)
	/home/xyj/Projects/tiflow/engine/enginepb/master_grpc.pb.go:433 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc001b50000, {0x5d7bfe0, 0xc001286ea0}, 0xc002066000, 0xc001b2edb0, 0x7f1c5a0, 0x0)
	/home/xyj/.local/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1295 +0xb2b
google.golang.org/grpc.(*Server).handleStream(0xc001b50000, {0x5d7bfe0, 0xc001286ea0}, 0xc002066000, 0x0)
	/home/xyj/.local/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:1636 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	/home/xyj/.local/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:932 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/home/xyj/.local/go/pkg/mod/google.golang.org/grpc@v1.48.0/server.go:930 +0x28a

```

### Versions of the cluster

master

--- a/engine/servermaster/scheduler/scheduler.go
+++ b/engine/servermaster/scheduler/scheduler.go
@@ -19,14 +19,14 @@ import (
 
 	"github.com/pingcap/tiflow/engine/model"
 	schedModel "github.com/pingcap/tiflow/engine/servermaster/scheduler/model"
+	"github.com/pingcap/tiflow/pkg/errors"
 )
 
 // Scheduler is a full set of scheduling management, containing capacity provider,
 // real scheduler and resource placement manager.
 type Scheduler struct {
-	infoProvider         executorInfoProvider
-	placementConstrainer PlacementConstrainer
-	filters              []filter
+	infoProvider executorInfoProvider
+	filters      []filter
 }
 
 // NewScheduler creates a new Scheduler instance
@@ -35,8 +35,7 @@ func NewScheduler(
 	placementConstrainer PlacementConstrainer,
 ) *Scheduler {
 	return &Scheduler{
-		infoProvider:         infoProvider,
-		placementConstrainer: placementConstrainer,
+		infoProvider: infoProvider,
 		filters: []filter{
 			newResourceFilter(placementConstrainer),
 			newSelectorFilter(infoProvider),
@@ -52,6 +51,9 @@ func (s *Scheduler) ScheduleTask(
 	candidates, err := s.chainFilter(ctx, request)
 	if err != nil {
 		return nil, err
+	}
+	if len(candidates) == 0 {
+		return nil, errors.ErrNoQualifiedExecutor.FastGenByArgs()
 	}
 	executorID := candidates[rand.Intn(len(candidates))]
 	return &schedModel.SchedulerResponse{ExecutorID: executorID}, nil

--- a/engine/servermaster/scheduler/scheduler.go
+++ b/engine/servermaster/scheduler/scheduler.go
@@ -53,7 +53,7 @@ func (s *Scheduler) ScheduleTask(
 		return nil, err
 	}
 	if len(candidates) == 0 {
-		return nil, errors.ErrNoQualifiedExecutor.FastGenByArgs()
+		return nil, errors.ErrNoQualifiedExecutor.GenWithStackByArgs()
 	}
 	executorID := candidates[rand.Intn(len(candidates))]
 	return &schedModel.SchedulerResponse{ExecutorID: executorID}, nil

--- a/engine/servermaster/scheduler/scheduler_test.go
+++ b/engine/servermaster/scheduler/scheduler_test.go
@@ -120,3 +120,13 @@ func TestSchedulerConstraintConflict(t *testing.T) {
 	require.Error(t, err)
 	require.True(t, errors.Is(err, errors.ErrResourceConflict))
 }
+
+func TestSchedulerNoQualifiedExecutor(t *testing.T) {
+	sched := NewScheduler(
+		&mockExecutorInfoProvider{infos: map[model.ExecutorID]schedModel.ExecutorInfo{}},
+		&MockPlacementConstrainer{},
+	)
+
+	_, err := sched.ScheduleTask(context.Background(), &schedModel.SchedulerRequest{})
+	require.ErrorIs(t, err, errors.ErrNoQualifiedExecutor)
+}

--- a/errors.toml
+++ b/errors.toml
@@ -1591,6 +1591,11 @@ error = '''
 metastore %s does not exist
 '''
 
+["DFLOW:ErrNoQualifiedExecutor"]
+error = '''
+no executor is available for scheduling
+'''
+
 ["DFLOW:ErrNoRPCClient"]
 error = '''
 no available RPC client

--- a/pkg/errors/engine_errors.go
+++ b/pkg/errors/engine_errors.go
@@ -351,6 +351,10 @@ var (
 		"local resource directory not writable",
 		errors.RFCCodeText("DFLOW:ErrLocalFileDirNotWritable"),
 	)
+	ErrNoQualifiedExecutor = errors.Normalize(
+		"no executor is available for scheduling",
+		errors.RFCCodeText("DFLOW:ErrNoQualifiedExecutor"),
+	)
 	ErrFilterNoResult = errors.Normalize(
 		"filter % returns no result",
 		errors.RFCCodeText("DFLOW:ErrFilterNoResult"),

